### PR TITLE
Deprecate VectorizedArray::n_array_elements

### PR DESCRIPTION
**Pull request by peterrum**
As discussed in https://github.com/dealii/dealii/pull/9681#issuecomment-601204517.

For some reason `matrix_free/matrix_vector_faces_28.debug` fails, but that also fails on master.

### CODE REVIEW

File: include/deal.II/base/vectorization.h [hunk 1]
```diff
@@ -502,7 +500,7 @@ class VectorizedArray
   }
 
   /**
-   * Load @p n_array_elements from memory into the calling class, starting at
+   * Load @p size() from memory into the calling class, starting at
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]: The old text was broken, but it got worse now. Suggestion:
```suggestion
   * Load @p size() data items from memory into the calling class, starting at
```

File: include/deal.II/base/vectorization.h [hunk 2]
```diff
@@ -516,7 +514,7 @@ class VectorizedArray
 
   /**
    * Write the content of the calling class into memory in form of @p
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * Write the content of the calling class into memory in form of
```

File: include/deal.II/base/vectorization.h [hunk 3]
```diff
@@ -516,7 +514,7 @@ class VectorizedArray
...
-   * n_array_elements to the given address. The pointer `ptr` needs not be
+   * size() to the given address. The pointer `ptr` needs not be
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * size() data items to the given address. The pointer `ptr` needs not be
```

File: include/deal.II/base/vectorization.h [hunk 4]
```diff
@@ -529,7 +527,7 @@ class VectorizedArray
 
   /**
    * Write the content of the calling class into memory in form of
-   * @p n_array_elements to the given address using non-temporal stores that
+   * @p size() to the given address using non-temporal stores that
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * size() data items to the given address using non-temporal stores that
```

File: include/deal.II/base/vectorization.h [hunk 5]
```diff
@@ -581,14 +579,14 @@ class VectorizedArray
   }
 
   /**
-   * Load @p n_array_elements from memory into the calling class, starting at
+   * Load @p size() from memory into the calling class, starting at
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * Load @p size() data items from memory into the calling class, starting at
```

File: include/deal.II/base/vectorization.h [hunk 6]
```diff
@@ -601,13 +599,13 @@ class VectorizedArray
 
   /**
    * Write the content of the calling class into memory in form of @p
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * Write the content of the calling class into memory in form of
```

File: include/deal.II/base/vectorization.h [hunk 7]
```diff
@@ -601,13 +599,13 @@ class VectorizedArray
...
-   * n_array_elements to the given address and the given offsets, filling the
+   * size() to the given address and the given offsets, filling the
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * size() data items to the given address and the given offsets, filling the
```

File: include/deal.II/base/vectorization.h [hunk 8]
```diff
@@ -750,13 +748,13 @@ inline DEAL_II_ALWAYS_INLINE VectorizedArrayType
 
 
 /**
- * Load @p n_array_elements from memory into the the VectorizedArray @p out,
+ * Load @p size() from memory into the the VectorizedArray @p out,
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
 * Load size() data items from memory into the the VectorizedArray @p out,
```

File: include/deal.II/base/vectorization.h [hunk 9]
```diff
@@ -1069,7 +1055,7 @@ class VectorizedArray<double, 8>
   }
 
   /**
-   * Load @p n_array_elements from memory into the calling class, starting at
+   * Load @p size() from memory into the calling class, starting at
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * Load size() data items from memory into the calling class, starting at
```
same at all occurrences below similarly to the main class.

File: include/deal.II/base/vectorization.h [hunk 10]
```diff
@@ -5280,7 +5265,7 @@ namespace std
   /**
    * Compute the sine of a vectorized data field. The result is returned as
    * vectorized array in the form <tt>{sin(x[0]), sin(x[1]), ...,
-   * sin(x[n_array_elements-1])}</tt>.
+   * sin(x[size()-1])}</tt>.
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h]:
```suggestion
   * sin(x[VectorizedArray::size()-1])}</tt>.
```
Here and below.

File: include/deal.II/lac/vector_operations_internal.h [hunk 11]
```diff
@@ -764,9 +764,8 @@ namespace internal
     template <typename Number, typename Number2>
     struct Dot
     {
-      static const bool vectorizes =
-        std::is_same<Number, Number2>::value &&
-        (VectorizedArray<Number>::n_array_elements > 1);
+      static const bool vectorizes = std::is_same<Number, Number2>::value &&
```
Comment (kronbichler) [inline include/deal.II/lac/vector_operations_internal.h]:
```suggestion
      static constexpr bool vectorizes = std::is_same<Number, Number2>::value &&
```
I am not 100% sure here if this requires an explicit instantiation on some compilers, similarly to what we do for `VectorizedArray::n_array_elements` - but it is worth a try.

File: include/deal.II/lac/vector_operations_internal.h [hunk 12]
```diff
@@ -1202,7 +1196,7 @@ namespace internal
       // First we work on (n_chunks/nvecs) chunks, where each chunk processes
       // nvecs*(4*8) elements.
 
-      const unsigned int nvecs = VectorizedArray<Number>::n_array_elements;
+      const unsigned int nvecs          = VectorizedArray<Number>::size();
```
Comment (kronbichler) [inline include/deal.II/lac/vector_operations_internal.h]:
```suggestion
      constexpr unsigned int nvecs          = VectorizedArray<Number>::size();
```

File: include/deal.II/lac/vector_operations_internal.h [hunk 13]
```diff
@@ -1221,8 +1215,7 @@ namespace internal
           r0 += r1;
           r2 += r3;
           r0 += r2;
-          r0.store(
-            &outer_results[i * VectorizedArray<Number>::n_array_elements]);
+          r0.store(&outer_results[i * VectorizedArray<Number>::size()]);
```
Comment (kronbichler) [inline include/deal.II/lac/vector_operations_internal.h]:
```suggestion
          r0.store(&outer_results[i * nvecs]);
```

File: include/deal.II/lac/vector_operations_internal.h [hunk 14]
```diff
@@ -1231,9 +1224,9 @@ namespace internal
       // regular_chunks * nvecs; We do as much as possible with 2 SIMD
       // operations within each chunk. Here we assume that nvecs < 32/2 = 16 as
       // well as 16%nvecs==0.
-      AssertIndexRange(VectorizedArray<Number>::n_array_elements, 17);
+      AssertIndexRange(VectorizedArray<Number>::size(), 17);
```
Comment (kronbichler) [inline include/deal.II/lac/vector_operations_internal.h]:
```suggestion
      static_assert(VectorizedArray<Number>::size() <= 16 &&
                    16 % VectorizedArray<Number>::size(),
                    "VectorizedArray::size() must be a power of 2 and not more than 16");
```
(for this and the next line)

File: include/deal.II/lac/vector_operations_internal.h [hunk 15]
```diff
@@ -1231,9 +1224,9 @@ namespace internal
...
       Assert(16 % nvecs == 0, ExcInternalError());
-      if (n_chunks % VectorizedArray<Number>::n_array_elements != 0)
+      if (n_chunks % VectorizedArray<Number>::size() != 0)
```
Comment (kronbichler) [inline include/deal.II/lac/vector_operations_internal.h]:
```suggestion
      if (n_chunks % nvecs != 0)
```

File: include/deal.II/matrix_free/evaluation_kernels.h [hunk 16]
```diff
@@ -2089,13 +2089,14 @@ namespace internal
         {
           AssertDimension(
             dof_info.n_vectorization_lanes_filled[dof_access_index][cell],
-            VectorizedArrayType::n_array_elements);
+            VectorizedArrayType::size());
```
Comment (kronbichler) [inline include/deal.II/matrix_free/evaluation_kernels.h]: I think this function (and the next one) would be a place to introduce an abbreviation like `n_simd = VectorizedArrayType::size()`. But that can also be done in another PR.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 17]
```diff
@@ -1758,7 +1755,7 @@ class FEEvaluationAccess<1, 1, Number, is_face, VectorizedArrayType>
  *         // Need to evaluate function for each component in VectorizedArray
  *         VectorizedArray<double> f_value;
  *         for (unsigned int v=0;
- *              v<VectorizedArray<double>::n_array_elements;
+ *              v<VectorizedArray<double>::size();
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: I think this for clause would fit in one line now.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 18]
```diff
@@ -3564,12 +3561,12 @@ template <int dim,
...
           typename VectorizedArrayType>
-std::array<unsigned int, VectorizedArrayType::n_array_elements>
+std::array<unsigned int, VectorizedArrayType::size()>
 FEEvaluationBase<dim, n_components_, Number, is_face, VectorizedArrayType>::
   get_cell_ids() const
 {
-  const unsigned int v_len = VectorizedArrayType::n_array_elements;
-  std::array<unsigned int, VectorizedArrayType::n_array_elements> cells;
+  const unsigned int v_len = VectorizedArrayType::size();
+  std::array<unsigned int, VectorizedArrayType::size()> cells;
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]:
```suggestion
  std::array<unsigned int, v_len> cells;
```

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 19]
```diff
@@ -3831,8 +3826,7 @@ FEEvaluationBase<dim, n_components_, Number, is_face, VectorizedArrayType>::
 
   // Case 3: standard operation with one index per degree of freedom -> go on
   // here
-  constexpr unsigned int n_vectorization =
-    VectorizedArrayType::n_array_elements;
+  constexpr unsigned int n_vectorization = VectorizedArrayType::size();
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: Looking through these changes I realize that we use lots of different abbreviations - at some point (not necessarily this PR) we should use a more uniform style for these short names; my favorite right now is `n_simd` or `n_lanes`.

File: include/deal.II/matrix_free/matrix_free.h [hunk 20]
```diff
@@ -2488,16 +2484,14 @@ MatrixFree<dim, Number, VectorizedArrayType>::n_active_entries_per_cell_batch(
...
   AssertIndexRange(cell_batch_number, task_info.cell_partition_data.back());
-  unsigned int n_components = VectorizedArrayType::n_array_elements;
-  while (
-    n_components > 1 &&
-    cell_level_index[cell_batch_number * VectorizedArrayType::n_array_elements +
-                     n_components - 1] ==
-      cell_level_index[cell_batch_number *
-                         VectorizedArrayType::n_array_elements +
-                       n_components - 2])
+  unsigned int n_components = VectorizedArrayType::size();
```
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.h]:
```suggestion
  unsigned int n_simd_filled = VectorizedArrayType::size();
```
and all downstream uses; I must admit that at first I was confused by the use of `n_components` here.

File: include/deal.II/matrix_free/matrix_free.h [hunk 21]
```diff
@@ -2509,12 +2503,12 @@ MatrixFree<dim, Number, VectorizedArrayType>::n_active_entries_per_face_batch(
   const unsigned int face_batch_number) const
 {
   AssertIndexRange(face_batch_number, face_info.faces.size());
-  unsigned int n_components = VectorizedArrayType::n_array_elements;
+  unsigned int n_components = VectorizedArrayType::size();
```
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.h]: Same here

File: include/deal.II/matrix_free/shape_info.h [hunk 22]
```diff
@@ -150,7 +150,7 @@ namespace internal
       /**
        * Stores the shape values of the 1D finite element evaluated on all 1D
        * quadrature points in vectorized format, i.e., as an array of
-       * VectorizedArray<dim>::n_array_elements equal elements. The length of
+       * VectorizedArray<dim>::size() equal elements. The length of
        * this array is <tt>n_dofs_1d * n_q_points_1d</tt> and quadrature
        * points are the index running fastest.
        */
```
Comment (kronbichler) [inline include/deal.II/matrix_free/shape_info.h]: This comment is actually outdated since we changed the data field to `Number` a while ago. Can you simply delete the statement starting from `in vectorized format` and replace it by `If the Number type is a VectorizedArray, the same value will be broadcast to all lanes.` Change here and in all other places. Note also that this will likely conflict with #9655 so we need to be careful here.

**Comment by kronbichler** [comment]
This deprecation should get an entry in the incompatibilities section of the changelog.

**Comment by kronbichler** [comment]
> For some reason `matrix_free/matrix_vector_faces_28.debug` fails, but that also fails on master.

This must be specific to your machine because it have not seen it fail on any of the results reported in cdash. In any case, the test is rather recent so we could dig into it in case it is necessary.

**Comment by peterrum** [comment]
> Thanks for this epic change.

I thought the change would be more tedious: a lot of warnings. But it turned out that replacing all instances is sufficient.

File: include/deal.II/base/vectorization.h [hunk 23]
```diff
@@ -1498,7 +1484,7 @@ vectorized_transpose_and_store(const bool                        add_into,
  */
 template <>
 class VectorizedArray<float, 16>
-  : public VectorizedArrayBase<VectorizedArray<float, 16>>
+  : public VectorizedArrayBase<VectorizedArray<float, 16>, 16>
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: Is this an unrelated change?

File: include/deal.II/base/vectorization.h [hunk 24]
```diff
@@ -245,18 +245,17 @@ class VectorizedArrayIterator
...
   /**
-   * Return the number of elements in the array stored in the variable
-   * n_array_elements of VectorizedArray.
+   * Return the number of elements in the array.
    */
   static constexpr std::size_t
   size()
   {
-    return T::n_array_elements;
+    return width;
```
Comment (peterrum) [inline include/deal.II/base/vectorization.h]: @bangerth Answering https://github.com/dealii/dealii/pull/9688#discussion_r395721182: this is the reason. We cannot access `n_array_elements` anymore.

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 17 above)
Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation.h]: Jup.

File: include/deal.II/matrix_free/shape_info.h (hunk 22 above)
Comment (peterrum) [reply include/deal.II/matrix_free/shape_info.h]: I have removed the changes form `shape_info` so that we don't get any merge issues.

**Comment by peterrum** [comment]
@kronbichler I have integrated most of your requests. I would suggest to name the variables consistently is the task of another PR.

**Comment by bangerth** [comment]
I spent a quality hour last night reading the "C++ Technical Specifications for Parallelism v2". They will probably be part of C++23. The specification contains a `std::simd` data type that does in essence what `VectorizedArray` does and, worth pointing out in the current context, has a `constexpr std::simd::size()` member function :-)
